src/magic_folder/snapshot.py

```diff
@@ -0,0 +1,172 @@
+
+@inlineCallbacks
+def tahoe_create_snapshot_dir(nodeurl, content, parents, timestamp, treq):
+    """
+    :param unicode content: readcap for the content.
+    :param [unicode] parents: List of parent snapshot caps
+    :param unicode author: readcap that represents the author pubkey
+    :param integer timestamp: POSIX timestamp that represents the creation time
+    :return Deferred[unicode]: The readcap associated with the newly created
+        snapshot.
+    """
+
+    # dict that would be serialized to JSON
+    body = \
+    {
+        u"content": [ "filenode", { u"ro_uri": content,
+                                    u"metadata": { } } ],
+        u"version": [ "filenode", { u"ro_uri": str(SNAPSHOT_VERSION),
+                                    u"metadata": { } } ],
+        u"timestamp": [ "filenode", { u"ro_uri": str(timestamp),
+                                      u"metadata": { } } ],
+    }
+
+    # populate parents
+    # The goal is to populate the dictionary with keys u"parent0", u"parent1" ...
+    # with corresponding dirnode values that point to the parent URIs.
+    if parents != []:
+        parent_nodes =  [ [ "dirnode", { u"ro_uri": p } ] for p in parents ]
+        for (k, v) in zip(count(0), parent_nodes):
+            body[unicode("parent" + str(k), 'utf-8')] = v
+
+    body_json = json.dumps(body)
+
+    node_url = DecodedURL.from_text(unicode(nodeurl, 'utf-8'))
+
+    # POST /uri?t=mkdir-immutable
+    url = node_url.child(
+        u"uri",
+    ).add(
+        u"t",
+        u"mkdir-immutable"
+    )
+
+    post_uri = url.to_uri().to_text().encode("ascii")
+    response = yield treq.post(post_uri, body_json)
+    if response.code != OK:
+        returnValue((yield bad_response(url, response)))
+
+    result = yield readBody(response)
+    returnValue(result)
+
+@inlineCallbacks
+def _store_file_immutable(nodeurl, filepath):
+
+    from twisted.internet import reactor
+    treq = HTTPClient(Agent(reactor))
+    rocap = yield tahoe_put_immutable(nodeurl, filepath, treq)
+
+    returnValue(rocap)
+
+@inlineCallbacks
+def snapshot_create(node_directory, filepath, parents):
+    """
+    Create a snapshot, given a file contents of a named file,
+    parent snapshots, author's identity and signature.
+
+    :param unicode filepath: The file path whose snapshot is being created
+
+    :param [unicode] parents: List of parent snapshots of the current snapshot
+        (read-caps of parent snapshots)
+
+    :return Deferred[unicode]: Snapshot read-only cap is returned on success.
+        Otherwise an appropriate exception is raised.
+    """
+
+    nodeurl = get_node_url(node_directory)
+
+    # - store the file content and get the immutable cap (content)
+    #     PUT /uri?format=CHK
+    content_cap = yield _store_file_immutable(nodeurl, filepath)
+
+    now = time.time()
+
+    # - HTTP POST mkdir-immutable
+    from twisted.internet import reactor
+    treq = HTTPClient(Agent(reactor))
```

Comment:
treq should be a parameter

Reply:
As a general point, I've usually tried to have `reactor` be an argument to functions that need it .. this at bare minimum lets you pass a `Clock()` (or whatever) for testing and avoids the imports inside a method. (Obviously here that is captured by "pass the http client itself" as @exarkun suggests). 

I guess I'm partly mentioning it here as a question: *should* this code-base prefer to have the reactor (or reactor-using thing) passed in? I think "yes".

Reply:
> I guess I'm partly mentioning it here as a question: should this code-base prefer to have the reactor (or reactor-using thing) passed in? 
I do not have knowledge about twisted so deeply to answer that question.